### ISSUE TITLE
ReadOnlySpan<char> overload of Uri.UnescapeDataString

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/runtime#54828]
## Background and Motivation

For context, this emerged in ASP.NET Core at https://github.com/dotnet/aspnetcore/issues/33840#issuecomment-869005822

Currently, `Uri.UnescapeDataString` both accepts and returns `string`. If the source data you're working with is a `ReadOnlySpan<char>`, then the caller is forced to allocate a `string`. This is especially inefficient if the caller only needs back a `ReadOnlySpan<char>`, since in most cases this seems like it should be achievable without any allocations (because the original argument could be returned directly if no transformation was required - see below).

## Proposed API

```diff
namespace System
{
     public class Uri
     {
+        ReadOnlySpan<char> UnescapeDataString(ReadOnlySpan<char> stringToUnescape) { ... }
     }
}
```

Of course, if the input data contains escaped characters, then the implementation will need to allocate in order to provide a return value. However the existing logic within `string UnescapeDataString(string)` recognizes the extremely common case where the input string does not contain escaped characters and knows to return the original `string` directly. Similarly, the new `ReadOnlySpan<char>` overload could return the original argument value in this case.

## Usage Examples

```cs
var unescapedValue = Uri.UnescapeDataString(someCharSpanFromAUri);
```

**Comment by GrabYourPitchforks** [comment]
We try to avoid APIs which allocate then return the allocated memory as a `ROS<T>` rather than as a concrete type, even if there's only a slight chance of it doing so. Most APIs that return values try to return them as concrete types since these are generally more usable.

One way to address this could be to return a `string?` instead of a `ROS<char>`, with a null return value meaning "the original data did not require transformation." It's a little bit weirder of an API shape since the input and output types are different, but it does provide fuller control for the caller. Alternatively, you could have the caller specify both an input and an output buffer, which is more common for transformation-style APIs.

**Comment by SteveSandersonMS** [comment]
Thanks @GrabYourPitchforks. I wasn't aware of the convention to avoid allocating and returning `ROS<T>` but I see how that does work against this suggestion.

> Alternatively, you could have the caller specify both an input and an output buffer, which is more common for transformation-style APIs.

TBH I don't think the caller would want to supply a buffer given that it's hard to predict the correct size (without an understanding of the decoding rules), and because it leads the caller into a complex world of needing to pool buffers etc.

> One way to address this could be to return a string? instead of a ROS<char>, with a null return value meaning "the original data did not require transformation."

It's slightly unusual, but I agree that does match the scenario. Perhaps the API would be something like:

```cs
bool DecodeIfRequired(ReadOnlySpan<char> value, [MaybeNullWhen(false)] out string result)
```

We're not blocked on this for 6.0 (at least, we're not in any way counting on this happening for 6.0!) but is potentially an opportunity for the future.

**Comment by Wraith2** [comment]
> ```csharp
> bool DecodeIfRequired(ReadOnlySpan<char> value, [MaybeNullWhen(false)] out string result)
> ```

wouldn't that more naturally fit with the try pattern?

```csharp
bool TryDecode(ReadOnlySpan<char> value, [MaybeNullWhen(false)] out string result)
```
or even/also
```csharp
(bool Success, string DecodedValue) TryDecode(ReadOnlySpan<char> value)
```

so usage would become:

```csharp
public void DoSomethingWithUri(ReadOnlySpan<char> urlChars)
{
    if (Uri.TryDecode(uriChars) is { Success: true, DecodedValue: var decoded })
    {
        uriChars = decoded;
    }
    // carry on and use uriChars
}
```

**Comment by Sergio0694** [comment]
I'll admit I'm not the biggest far of value-tuples in public APIs in general, but still I'm not sure I get how would:

```csharp
if (Uri.TryDecode(uriChars) is { Success: true, DecodedValue: var decoded })
{
}
```

be in any way better than:

```csharp
if (Uri.TryDecode(uriChars, out var decoded)
{
}
```

Isn't it just more verbose and awkward compared to the usual `TryFoo` pattern? 🤔

**Comment by Wraith2** [comment]
> Isn't it just more verbose and awkward compared to the usual TryFoo pattern?

I think so, but I also prefer the purity of separate input/output that using a value tuple allows. If they're going to be added I think the out param one is most likely with the valuetuple one a possibility but I wanted to make sure it was there for consideration.

**Comment by tannergooding** [comment]
CC. @karelz, this is another API proposal that impacts `Uri`. It's likely better handled by the networking team where the domain expertise resides.

**Comment by karelz** [comment]
Looks like a need for span overload -- @MihaZupan any thoughts?

**Comment by karelz** [comment]
@tannergooding System.Net is the area for System.Uri -- it is called out in [area-owners](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) - feel free to move it area-System.Net in future. Thanks!

**Comment by MihaZupan** [comment]
Duplicate of #40603
